### ISSUE TITLE
Alt text in html and pdf document output

### ISSUE THREAD

**Issue by cmichaud92** [yihui/rmarkdown-cookbook#242]
Please consider adding a section on implementing accessibility features with rmarkdown.  I feel this is important functionality, however there is presently a lack of information (the web). I have hit a dead end trying to add alt text to figures generated in rmarkdown or imported into the document.  To be clear, I am in need of text which will not be visible to a sighted reader (as it would clutter the document) but will be read by a line reader.  I am assuming this is possible in rmarkdown.  This is a very small thing which will be of great value to people with visual disabilities.  I appreciate the consideration.

**Comment by cderv** [comment]
Personally, I am do not know how this work in html. If you can point to any resource regarding how to do that for HTML, I can have a look to see how it could work with Rmarkdown and knitr. 

However, maybe one of my co-author will already know this.

**Comment by emilyriederer** [comment]
I researched the HTML side of this a bit to send a similar PR for `blastula` (https://github.com/rich-iannone/blastula/pull/40) . For raw HTML, it's a matter of adding the `alt` attribute to the `img` tag. 

I actually was researching this recently to try to make an accessible document for UseR! If my memory serves, I think I recall reading somewhere online that one can use `fig.cap` chunk option but (optionally) suppress the caption actually printing with `fig_caption: false` in YAML. However, I ended up doing that task a different way, so I would have to check to confirm that that actually works. 

I'm least certain for LaTeX.

**Comment by cmichaud92** [comment]
Emily,  I was able to get `fig_caption: false` to suppress printing the caption and capture the text as alt text in the output.  I had tried this in the past, I was short an indent in the yaml, it would fail silently, devil is in the details I guess.  This is great, however, alt text is a rich description of the visual (not a one liner), this is a bit awkward to cram into the chunk header .  

Here are a couple of r related accessibility resources.  The only ones I could locate following some effort
[Accessible R Marksown Documents](https://r-resources.massey.ac.nz/rmarkdown/) gives a sparse overview.  I'm particularly interested in  JooYoung Seo's [post](https://jooyoungseo.com/post/ds4blind/) on using sonify and especially BrailleR packages to produce accessable R scripts.  To be honest I have not yet tried to implement these in R markdown as I fear a very large time commitment (I'm a biologist, not a software engineer), but the benefits would be substantial.  Thanks for your and cderv's time

**Comment by cderv** [comment]
So that means using `fig.cap` to create a <figcaption> should work ? I mean screen reader can handle  this tag right ? 

Pandoc already deal with figure this way when converting from markdown to HTML
```bash
echo '![foo](bar.png)' | pandoc
<figure>
<img src="bar.png" alt="" /><figcaption>foo</figcaption>
</figure>
```
I think knitr will convert by default the figures to the markdown syntax for pandoc to convert to html like above. 
Like with 
````md
---
output:  html_document:
---

```{r, fig.cap = "a plot of mtcars"}
plot(mtcars)
```
````

If you trick knitr by adding attributes, like `fig.align`, you would get the `alt` text
From 
````md
```{r, fig.cap = "a plot of mtcars", fig.align='center'}
plot(mtcars)
```
````
to
````html
<div class="figure" style="text-align: center">
<img src="test_files/figure-html/unnamed-chunk-1-1.png" alt="a plot of mtcars"  />
<p class="caption">a plot of mtcars</p>
</div>
````

about `fig_caption` in `html_document`, it controls the Pandoc extension [implicit_figures](https://pandoc.org/MANUAL.html#extension-implicit_figures). It will change the way pandoc converts the markdown string
````bash
echo '![foo](bar.png)' | pandoc -f markdown-implicit_figures
<p><img src="bar.png" alt="foo" /></p>
````

I think there is room to rethink how all this work (or at least clarify it) to make it more usable for accessibility.

**Comment by cderv** [comment]
> however, alt text is a rich description of the visual (not a one liner), this is a bit awkward to cram into the chunk header .

You can pass a variable that contains your text in the chunk options

````md
---
output: 
  html_document:
    fig_caption: false
    keep_md: true
---

```{r, include = FALSE}
long_fig_cap <- "This is a long figcaption
that is on several lines. But also with 
several sentences."
```


```{r, fig.cap = long_fig_cap}
plot(mtcars)
```
````

Or even put the alt text in a file and read it by using `xfun::file_string()` function
````md
---
output: 
  html_document:
    fig_caption: false
    keep_md: true
---

```{r, fig.cap = xfun::file_string("alt.txt")}
plot(mtcars)
```
```` 
with a file `alt.txt` containing your text. 

You can execute any R code as chunk option, so you could even find other system.

**Comment by yihui** [comment]
I think this question was also recently asked in RStudio Community: https://community.rstudio.com/t/70722

**Comment by cmichaud92** [comment]
cderv, Thanks for this solution.  I really appreciate it.

> You can pass a variable that contains your text in the chunk options

Yihui, This question was also  asked [here](https://community.rstudio.com/t/adding-alt-text-to-figures-generated-in-r-markdown/65191) (by me).  I was interested in HanOostdijk solution to the question you linked, however the resulting output seems off.  The alt text is tagged as 'title': `<img src="data:..." title="my alt text" alt> == $0`.  This might be fine (I know little about html) and am still learning about line readers. When I run the code suggested by cderv the alt text is tagged as alt text`<img src="data:..." alt="my alt text" > == $0`.   *I've been looking for this exact solution for over a month*. 

Myself and an increasing number of researchers would really benefit from an accessibility resource.  It is a priority of mine to expand my audience. A very simple step toward this goal is by implementing accessibility measures.  i feel that this book would be  an ideal place for this information.  Thanks again.

**Comment by yihui** [comment]
I'll consider adding a section to address the accessibility issues, although I'm not an expert in this area. Pointing to https://r-resources.massey.ac.nz/rmarkdown/ might be helpful enough (I've met its author in 2014 and he showed me exactly the `alt` problem).

**Comment by yihui** [comment]
@cmichaud92 Regarding the empty `alt` attribute, I just posted a reply at https://community.rstudio.com/t/70722/14?u=yihui

**Comment by yihui** [comment]
I just added a section at https://bookdown.org/yihui/rmarkdown-cookbook/html-accessibility.html. Feedback welcome!
